### PR TITLE
fix(providers): collect BlockState before constructing DB provider

### PR DESCRIPTION
**Pull request by Rjected**
Changes from https://github.com/paradigmxyz/reth/pull/11265 that fix a race condition in the builder

The race condition is between the builder and tree, with the following order of events:
* The builder is trying to get a state provider for a block, with anchor = block `N`
* The persistence task finishes, and cleans up block `N+1` from the in memory state
* The builder fetches calls `CanonicalInmemoryState::state_provider`. In the current code this results in a DB provider for block `N`, and `BlockState` for `N+2, ...`, meaning there is a gap for block `N+1`, which no longer exists in `CanonicalInMemoryState`.

This happens because the anchor hash is determined before block `N+1` is persisted. The `BlockState` is still valid, but we should not be re-fetching the state.

This PR re-uses the `BlockState` that is used to determine the anchor hash, and introduces a new method `state_provider_from_state` that allows this reuse.

### CODE REVIEW

**Comment by rkrasiuk** [comment]
tyty, will approve after trace cleanup

File: crates/storage/provider/src/providers/blockchain_provider.rs [hunk 1]
```diff
@@ -175,9 +175,12 @@ impl<N: ProviderNodeTypes> BlockchainProvider2<N> {
         state: impl AsRef<BlockState>,
     ) -> ProviderResult<MemoryOverlayStateProvider> {
         let state = state.as_ref();
+        trace!(target: "providers::blockchain", "Calling anchor for fetching historical state provider");
         let anchor_hash = state.anchor().hash;
+        trace!(target: "providers::blockchain", ?anchor_hash, "Fetching historical provider for block hash");
         let latest_historical = self.database.history_by_block_hash(anchor_hash)?;
```
Comment (rkrasiuk) [inline crates/storage/provider/src/providers/blockchain_provider.rs:181]: can there still be a race condition here between the moment anchor hash was set on `BlockState` and the moment when we actually retrieve the state provider? does it matter here?

Comment (Rjected) [reply crates/storage/provider/src/providers/blockchain_provider.rs:181]: >  can there still be a race condition here between the moment anchor hash was set on BlockState and the moment when we actually retrieve the state provider? does it matter here?

I think so, although only for reorgs afaict. When the DB state only moves forward this should not be a problem. Take for example a reorg which removes one block from the DB. The race is:
* we fetch the state
* anchor is removed from DB, replaced with another block
* now we call `history_by_block_hash` on something that does not exist

The only idea I can come up with to fix it, is to store `ExecutedBlock`s like we do `TrieUpdates` before finalization.

**Comment by Rjected** [comment]
cc @rkrasiuk traces removed

File: crates/storage/provider/src/providers/blockchain_provider.rs [hunk 2]
```diff
@@ -177,7 +177,7 @@ impl<N: ProviderNodeTypes> BlockchainProvider2<N> {
         let state = state.as_ref();
         let anchor_hash = state.anchor().hash;
         let latest_historical = self.database.history_by_block_hash(anchor_hash)?;
-        Ok(self.canonical_in_memory_state.state_provider(state.hash(), latest_historical))
+        Ok(self.canonical_in_memory_state.state_provider_from_state(state, latest_historical))
```
Comment (mattsse) [inline crates/storage/provider/src/providers/blockchain_provider.rs:180]: ah massive yikes

File: crates/chain-state/src/in_memory.rs [hunk 3]
```diff
@@ -494,6 +494,21 @@ impl CanonicalInMemoryState {
...
+    /// Return state provider with reference to in-memory blocks that overlay database state.
+    ///
+    /// This merges the state of all blocks that are part of the chain that the requested block is
+    /// the head of. This includes all blocks that connect back to the canonical block on disk.
+    pub fn state_provider_from_state(
+        &self,
+        state: &BlockState,
+        historical: StateProviderBox,
+    ) -> MemoryOverlayStateProvider {
+        let in_memory: Vec<_> =
```
Comment (mattsse) [inline crates/chain-state/src/in_memory.rs:506]:
```suggestion
        let in_memory =
```

**Review by mattsse — APPROVED** [review]
uffff

**Pull request merged**